File: lib/dao.js
/**
 * @author    hc
 * @time      2019-10-18
 * @change    2020-04-26
 * 数据库持久层
 *    1， 提供操作数据库的方法
 *    2， 防止sql注入攻击
 *    3,  如果数据库不存在则初始化数据库
 *
 *     记得归还连接    ----------    你还欠我一个连接
 *
 *
 *     数据库备份
 *     mysqldump -h主机名 -P端口 -u用户名 -p密码 --database 数据库名 > 文件名.sql
 */

const mysql = require('mysql');
const config = require('./config');
const logger = require('./logger');
const dbs = require('./dbs.js');
const tables = dbs.config;

// 创建数据库连接
const cnnPool = mysql.createPool({
  host: config.dbHost || 'localhost',
  port: config.dbPort || '3306',
  user: config.dbUser || 'root',
  password: config.dbPasswd || '123',
  database: config.dbDatabase || 'bysj', // 优先使用数据库配置中的数据库名 而不是全局的数据库名
  connectionLimit: config.dbPool || 10,
  connectTimeout: config.dbTimeOut || 30,
  multipleStatements: config.dbMultiSQL || false,
  stringifyObjects: true, // 序列化对象
  typeCast: true, // 转换列值为本地JavaScript数据类型
});

// 防止连接池初始化完成之前获取连接， 这种做法非常hack不保险，后期考虑修改
setTimeout(() => checkCnnPool(), 300);

/**
 * 测试数据库连接池
 */
function checkCnnPool() {
  logger.info('数据库自检中...');
  const sql = 'SELECT 1 + 1 AS result;';
  myQuery(sql, (data) => {
    if (data === false) {
      logger.error('数据库连接池初始化失败');
      process.exit();
    } else {
      logger.info('数据库连接池初始化成功');
      checkDatabase();
    }
  });
  process.once('uncaughtException', processEixt);
  process.once('beforeExit', processEixt);
  process.once('exit', processEixt);
  process.once('SIGINT', processEixt);
  function processEixt() {
    logger.info(`关闭数据库`);
    cnnPool.end(); // 确保所有的查询sql语句执行完成之后再退出
    process.exit();
  }
}

/**
 * 数据库自检
 */
function checkDatabase() {
  const keys = Object.keys(dbs.items);
  const lis = [];
  keys.forEach((key) => {
    lis.push(
      new Promise((res, rej) => {
        const items = dbs.items[key];
        logger.info(`检测${items.msg}...`);
        myQuery(items.check, (data) => {
          if (data && data.length <= 0) {
            myQuery(items.init, (data) => {
              if (data) {
                logger.info(`${items.msg}不存在，已自动构建...`);
                res(data);
              } else {
                logger.fatal(`${items.msg}不存在，构建失败...`);
                rej(false);
              }
            });
          }
          res(data);
        });
      })
    );
  });
  Promise.all(lis).then((res) => {
    if (res.every((item) => item && item.length > 0)) {
      logger.info('数据库初始化完成...');
    }
  });
}

/**
 * 简单封装的查询语句，自动从连接池里获取连接，处理程序异常，完成查询之后自动放回连接池
 * @param {*} String  SQL语句
 * @param {*} callback 回调函数
 */
function myQuery(sql, callback) {
  cnnPool.getConnection((error, cnn) => {
    if (error) return callback(false);
    cnn.query(sql, (error, data, feild) => {
      if (error) {
        logger.error(error);
        callback(false);
      } else {
        callback(data);
      }
    });
    cnn.release();
  });
}

module.exports = {
  /**
   * 通过关键字模糊查询
   * @param {*} kw 关键字
   * @param {*} bg 起始位置
   * @param {*} len 结束位置
   * @param {*} callback 会掉函数
   */
  getByKeyWord(kw, bg, len, callback) {
    const sql = `SELECT  *  FROM ${tables.table_image} WHERE TAG LIKE '%${kw}%' limit ${bg},${len};`;
    myQuery(sql, callback);
  },

  /**
   * 通过关键字查询
   * @param {*} kw 关键字
   * @param {*} callback 回调函数
   */
  getCountByKeyWord(kw, callback) {
    const sql = `SELECT COUNT(*) AS total FROM ${tables.table_image} WHERE TAG LIKE '%" + kw + "%';`;
    myQuery(sql, callback);
  },

  /**
   * 从图片表中查找所有的图片分类
   * @param {*} callback
   */
  getAllImageType(callback) {
    const sql = `SELECT distinct category FROM ${tables.table_image}`;
    myQuery(sql, callback);
  },

  /**
   * 获取指定图片类型的数据
   * @param {String} category 分类信息
   * @param {number} count  便宜量
   * @param {number} len  长度
   * @param {Function} callback  回调函数
   */
  getImageByCategory(category, count, len, callback) {
    const sql = `SELECT  *  FROM ${tables.table_image} WHERE category='${category}' limit ${count},${len};`;
    myQuery(sql, callback);
  },

  /**
   * 获取指定分类图片 的数量
   * @param {String} category 分类信息
   * @param {Function} callback  回调函数
   */
  getImageLenByCategory(category, callback) {
    const sql = `SELECT COUNT(*) AS total FROM ${tables.table_image} WHERE category='${category}';`;
    myQuery(sql, callback);
  },

  /**
   * 获取必应最近7日的数据
   * @param {Function} callback
   */
  getBiyingServenDay(callback) {
    const sql = `SELECT * FROM ${tables.table_biying} ORDER BY date desc limit 0,7;`;
    myQuery(sql, callback);
  },

  /**
   * 获取最新金山词典数据
   * @param {Function} callback
   */
  getEnglishDaily(callback) {
    const sql = `SELECT * FROM ${tables.table_english} ORDER BY time DESC LIMIT 0,1;`;
    myQuery(sql, callback);
  },

  /**
   * 账户登录的用户检查
   * @param {String} count
   * @param {Strign} passwd
   * @param {Function} callback
   */
  getUserByCount(count, passwd, callback) {
    const sql = `SELECT * FROM ${tables.table_user} WHERE count='${count}' and passwd='${passwd}';`;
    myQuery(sql, callback);
  },

  /**
   * 检查指定用户名是否存在
   * @param {String} count
   * @param {Function} callback
   */
  getUserExistsByCount(count, callback) {
    const sql = `SELECT * FROM ${tables.table_user} WHERE count='${count}';`;
    myQuery(sql, callback);
  },

  /**
   * 添加用户
   * @param {String} count
   * @param {String} passwd
   * @param {Function} callback
   */
  addUserByCount(count, passwd, callback) {
    const name = 'user' + new Date().getTime();
    const sql = `INSERT INTO ${tables.table_user} (count,passwd,name) values('${count}', '${passwd}', '${name}');`;
    myQuery(sql, callback);
  },

  /**
   * 获取用户信息
   * @param {Number} id 解密token的用户id
   * @param {Function} callback
   */
  getUserInfoAll(id, callback) {
    const sql = `SELECT * FROM ${tables.table_user} WHERE id='${id}';`;
    myQuery(sql, callback);
  },

  // 废弃， 校验图片相似度
  getImageByIdentify(identify, callback) {
    const sql = `SELECT * FROM ${tables.table_image} WHERE gary64='${identify}';`;
    console.log(sql);
    myQuery(sql, callback);
  },

  /**
   * 获取用户信息
   * @param {number} count
   * @param {number} len
   * @param {Function} callback
   */
  getAllUserInfo(count, len, callback) {
    const sql = `SELECT * FROM ${tables.table_user} LIMIT ${count},${len};`;
    myQuery(sql, callback);
  },
  /**
   * 删除用户信息
   * @param {number} id
   * @param {Function} callback
   */
  deleteUserInfo(id, callback) {
    const sql = `DELETE FROM ${tables.table_user} WHERE id='${id}';`;
    myQuery(sql, callback);
  },
};

File: lib/dao.py
'''
    @author    hec9527
    @time      2020-1-16
    @change    2020-1-16
    @description

    DB模块
    操作数据库，爬虫模块爬取到的数据并且存到数据库当中
'''

import os
import json
import pymysql.cursors
from sys import exit


class DBS:
    def __init__(self, **args):
        logger.info("数据库模块启动...")
        # load config
        config_path = os.path.abspath(
            os.path.join(os.path.dirname(__file__), "../config/config.json"))
        # load config
        with open(config_path) as file:
            self.config = json.load(file)
        # init cnn
        try:
            self.cnn = pymysql.connect(
                host=self.config.get("db_host", ''),
                user=self.config.get("db_user", ''),
                password=self.config.get("db_passwd", ''),
                db=self.config.get("db_database", ""),
                charset=self.config.get("db_charset", ''),
                cursorclass=pymysql.cursors.DictCursor)
        except Exception:
            logger.error('数据库连接初始化失败，请检查配置文件')
            exit()
        # check if databases is exist
        self.checkDB()

    # 检查数据库
    def checkDB(self):
        logger.info("数据库检查...")
        try:
            # 检查数据库是否存在
            sql = "SELECT * FROM information_schema.SCHEMATA where SCHEMA_NAME='bysj';"
            if self.myQuery(sql) is None:
                sql = "create database if not exists bysj;"
                self.myQuery(sql)
                logger.warning("数据库不存在，自动构建...")

            # 检查图片表是否存在
            sql = "SELECT table_name FROM information_schema.TABLES WHERE table_name ='images';"
            if self.myQuery(sql) is None:
                sql = "CREATE TABLE IF NOT EXISTS images(id INT(20) NOT NULL AUTO_INCREMENT PRIMARY KEY, url varchar(256) NOT NULL, resolution varchar(15), category varchar(50), tag varchar(256)), gary64 char(64));"
                self.myQuery(sql)
                logger.warning("spider表不存在，自动构建...")

            # 检查必应每日表是否存在
            sql = "SELECT table_name FROM information_schema.TABLES WHERE table_name ='biying';"
            if self.myQuery(sql) is None:
                sql = "CREATE TABLE IF NOT EXISTS biying(id INT(20) NOT NULL AUTO_INCREMENT PRIMARY KEY, url varchar(256) NOT NULL, title varchar(256), date char(10));"
                self.myQuery(sql)
                logger.warning("biying表不存在，自动构建...")

            # 检查每日英语表是否存在
            sql = "SELECT table_name FROM information_schema.TABLES WHERE table_name ='jinshan';"
            if self.myQuery(sql) is None:
                sql = "CREATE TABLE IF NOT EXISTS jinshan(id INT(20) NOT NULL AUTO_INCREMENT PRIMARY KEY, url varchar(256) NOT NULL, time char(10), english varchar(256), translate varchar(256), voice varchar(256));"
                self.myQuery(sql)
                logger.warning("jinshan表不存在，自动构建...")

            # 检查用户表是否存在
            # TODO 用户表增加 权限   0 表示管理员  非 0 表示普通用户
            sql = "SELECT table_name FROM information_schema.TABLES WHERE table_name ='uesrs';"
            if self.myQuery(sql) is None:
                '''create table if not exists user(id int(20) not null AUTO_INCREMENT PRIMARY KEY, count varchar(18) not null unique, passwd 
                    char(32) not null, permission int(1) default 3, name varchar(20) not null unique, avator varchar(256), age int(3), sex int(1), addr varchar(50), motto varchar(8
                    0), qq varchar(13), wechat varchar(20), email varchar(20), phone int(11), level int(3), vip int(3), balance varchar(10), lasttime 
                    timestamp, forbidden varchar(20));
                    '''
                sql = "CREATE TABLE IF NOT EXISTS user(id INT(20) NOT NULL AUTO_INCREMENT PRIMARY KEY); "
                self.myQuery(sql)
                logger.warning("users表不存在，自动构建...")
        except Exception:
            logger.critical("数据库初始化错误，请检查数据库配置")
            exit()

    # 插入数据到图片表
    def insertImage(self, url, tag, reso, category):
        sql = f"insert into images(url, category, tag, resolution) values('{url}', '{category}', '{tag}', '{reso}');"
        return self.myQuery(sql)

    # 插入数据到biying表
    def insertBiying(self, url, date, title):
        sql = f"INSERT INTO biying(url, date, title) values('{url}', '{date}', '{title}');"
        return self.myQuery(sql)

    # 从必应表搜索数据
    def selectFromBiying(self, date):
        sql = f"SELECT * FROM biying WHERE date='{date}';"
        return self.myQuery(sql)

    # 插入数据到金山词典
    def insertJinShan(self, time, url, english, translate, voice):
        sql = f"INSERT INTO jinshan(time, url, english, translate, voice) values('{time}', '{url}', '{english}', '{translate}', '{voice}');"
        return self.myQuery(sql)

    # 从金山词典中检查数据
    def selectFromJinShan(self, time):
        sql = f"SELECT * FROM jinshan WHERE time='{time}';"
        return self.myQuery(sql)

    # 封装的查询方法，减少重复代码的编写
    def myQuery(self, sql):
        try:
            cur = self.cnn.cursor()
            cur.execute(sql)
            self.cnn.commit()
            return cur.fetchone()
        except Exception as e:
            logger.error(e)
            return False
        finally:
            cur.close()

    # 释放链接
    def __del__(self):
        self.cnn.close()
        logger.info('数据库连接关闭')


# 开发测试
if __name__ == "__main__":
    db = DBS()
    from lib.logger import logger
else:
    from lib.logger import logger
